### PR TITLE
Update result status after being executed during EndBlock

### LINKED ISSUE #1046: Update result status after being executed during EndBlock

### PR DESCRIPTION
**Pull request by Benzbeeb**
Fixed #1046

### CODE REVIEW

File: chain/x/zoracle/internal/types/events.go [hunk 1]
```diff
@@ -10,6 +10,7 @@ const (
 	EventTypeReport              = "report"
 	EventTypeAddOracleAddress    = "add_oracle_address"
 	EventTypeRemoveOracleAddress = "remove_oracle_address"
+	EventTypeEndBlock            = "end_block"
```
Comment (sorawit) [inline chain/x/zoracle/internal/types/events.go]: EventTypeRequestExecute

**Review by sorawit — CHANGES_REQUESTED** [review]

File: chain/app/db_app.go [hunk 2]
```diff
@@ -202,7 +202,18 @@ func (app *dbBandApp) EndBlock(req abci.RequestEndBlock) (res abci.ResponseEndBl
 	if err != nil {
 		panic(err)
 	}
-	// Do other logic
+
+	events := res.GetEvents()
+	kvMap := make(map[string]string)
+	for _, event := range events {
+		if event.Type == zoracle.EventTypeRequestExecute {
```
Comment (taobun) [inline chain/app/db_app.go:247]: Process event one by one

**Review by taobun — CHANGES_REQUESTED** [review]

File: chain/x/zoracle/endblock.go [hunk 3]
```diff
@@ -16,22 +17,35 @@ func addUint64Overflow(a, b uint64) (uint64, bool) {
 	return a + b, false
 }
 
+func newRequestExecute(requestID RequestID, resolveStatus types.ResolveStatus) sdk.Event {
+	newAttributes := []sdk.Attribute{
```
Comment (sorawit) [inline chain/x/zoracle/endblock.go]: don't think we need this local variables.

File: chain/x/zoracle/endblock.go [hunk 4]
```diff
@@ -16,22 +17,35 @@ func addUint64Overflow(a, b uint64) (uint64, bool) {
 	return a + b, false
 }
 
+func newRequestExecute(requestID RequestID, resolveStatus types.ResolveStatus) sdk.Event {
```
Comment (sorawit) [inline chain/x/zoracle/endblock.go]: If we have this function. It should be named `newRequestExecuteEvent`

File: chain/x/zoracle/internal/types/request.go [hunk 5]
```diff
@@ -4,12 +4,12 @@ import (
 	sdk "github.com/cosmos/cosmos-sdk/types"
 )
 
-type ResolveStatus int8
+type ResolveStatus string
```
Comment (sorawit) [inline chain/x/zoracle/internal/types/request.go]: I don't think it's a good idea to save string to store. `uint8` is good as is. If you want event to be string, then create a function that converts enum to string.

**Review by sorawit — CHANGES_REQUESTED** [review]

File: chain/x/zoracle/endblock.go [hunk 6]
```diff
@@ -17,14 +17,13 @@ func addUint64Overflow(a, b uint64) (uint64, bool) {
...
-func newRequestExecute(requestID RequestID, resolveStatus types.ResolveStatus) sdk.Event {
-	newAttributes := []sdk.Attribute{
-		sdk.NewAttribute(types.AttributeKeyRequestID, fmt.Sprint(requestID)),
-		sdk.NewAttribute(types.AttributeKeyResolveStatus, string(resolveStatus)),
-	}
+func newRequestExecuteEvent(requestID RequestID, resolveStatus types.ResolveStatus) sdk.Event {
 	return sdk.NewEvent(
 		types.EventTypeRequestExecute,
-		newAttributes...,
+		[]sdk.Attribute{
```
Comment (sorawit) [inline chain/x/zoracle/endblock.go]: Why do we need create an array and spread it out?

File: chain/x/zoracle/internal/types/request.go [hunk 7]
```diff
@@ -4,14 +4,27 @@ import (
...
+func (resolveStatus ResolveStatus) String() string {
+	switch resolveStatus {
+	case Open:
+		return "Pending"
+	case Success:
+		return "Success"
+	case Failure:
+		return "Failure"
+	default:
+		return ""
```
Comment (sorawit) [inline chain/x/zoracle/internal/types/request.go]: Let's return unknown.

File: chain/x/zoracle/internal/types/request.go [hunk 8]
```diff
@@ -4,14 +4,27 @@ import (
...
-	Failure = ResolveStatus("Failure")
+	Open ResolveStatus = iota
+	Success
+	Failure
 )
 
+func (resolveStatus ResolveStatus) String() string {
+	switch resolveStatus {
+	case Open:
+		return "Pending"
```
Comment (sorawit) [inline chain/x/zoracle/internal/types/request.go]: Here let's just use Open?

**Review by sorawit — CHANGES_REQUESTED** [review]

File: chain/db/db.go [hunk 9]
```diff
@@ -450,3 +450,9 @@ func (b *BandDB) GetInvolvedAccountsFromTransferEvents(logs sdk.ABCIMessageLogs)
 	}
 	return involvedAccounts
 }
+
+func (b *BandDB) HandleEndBlock(id int64, resolveStatus string) error {
```
Comment (taobun) [inline chain/db/db.go]: Set result if resolveStatus is "SUCCESS"

File: chain/db/db.go [hunk 10]
```diff
@@ -450,3 +450,9 @@ func (b *BandDB) GetInvolvedAccountsFromTransferEvents(logs sdk.ABCIMessageLogs)
...
+	return b.tx.Where(Request{ID: id}).
+		Assign(Request{ResolveStatus: resolveStatus}).
```
Comment (taobun) [inline chain/db/db.go]: Use update instead, it must have request in table

File: chain/app/db_app.go [hunk 11]
```diff
@@ -241,7 +242,26 @@ func (app *dbBandApp) EndBlock(req abci.RequestEndBlock) (res abci.ResponseEndBl
...
+				if string(kv.Key) == zoracle.AttributeKeyRequestID {
+					requestID, err = strconv.ParseInt(string(kv.Value), 10, 64)
+					if err != nil {
+						break
+					}
+				} else if string(kv.Key) == zoracle.AttributeKeyResolveStatus {
+					resolveStatus = string(kv.Value)
+				}
+			}
+			app.dbBand.HandleEndBlock(requestID, resolveStatus)
```
Comment (taobun) [inline chain/app/db_app.go]: panic if err occured

**Review by taobun — CHANGES_REQUESTED** [review]

File: chain/app/db_app.go (hunk 11 above)
Comment (taobun) [reply chain/app/db_app.go]: Rename this function to `ResolveRequest`

**Review by taobun — CHANGES_REQUESTED** [review]

File: chain/app/db_app.go [hunk 12]
```diff
@@ -241,7 +242,32 @@ func (app *dbBandApp) EndBlock(req abci.RequestEndBlock) (res abci.ResponseEndBl
...
+	for _, event := range events {
+		if event.Type == zoracle.EventTypeRequestExecute {
+			var requestID int64
+			var resolveStatus string
+			var result []byte
+			for _, kv := range event.Attributes {
+				if string(kv.Key) == zoracle.AttributeKeyRequestID {
+					requestID, err = strconv.ParseInt(string(kv.Value), 10, 64)
+					if err != nil {
+						break
```
Comment (taobun) [inline chain/app/db_app.go]: If you serious about parse requestID should use `panic`, but if you want to ignore this request it should be `continue`

In this case, my opinion is using `panic`

**Review by taobun — CHANGES_REQUESTED** [review]

**Review by taobun — DISMISSED** [review]

File: chain/db/db.go [hunk 13]
```diff
@@ -462,3 +462,12 @@ func (b *BandDB) GetInvolvedAccountsFromTransferEvents(logs sdk.ABCIMessageLogs)
 	}
 	return involvedAccounts
 }
+
+func (b *BandDB) ResolveRequest(id int64, resolveStatus string, result []byte) error {
+	if resolveStatus == "Success" {
```
Comment (sorawit) [inline chain/db/db.go]: Let's expose `Success` in alias, and use `Success.String()` here.

**Review by sorawit — CHANGES_REQUESTED** [review]

**Review by sorawit — APPROVED** [review]

**Review by taobun — APPROVED** [review]

**Pull request merged**